Update README (#288)
gnoswap-labs/gnoswap: README.md

# GnoSwap Contracts

This repository contains smart contracts (realms) for GnoSwap.

## Index

- [Setting Up and Testing GnoSwap Contracts](#setting-up-and-testing-gnoswap-contracts)
  - [Prerequisites](#prerequisites)
  - [Setting Up GnoSwap Contracts](#setting-up-gnoswap-contracts)
  - [Running Tests](#running-tests)
- [Realms](#realms)
    - [Core Realms Deployed on Testnet4](#core-realms-deployed-on-testnet4)
    - [Pool](#pool)
    - [Position](#position)
    - [Router](#router)
    - [Staker](#staker)

## Setting Up and Testing GnoSwap Contracts

### Prerequisites

> - Latest version of Gno

Reviewer: ```suggestion
- Latest version of [gno.land](https://github.com/gnolang/gno)
```

This section guides you through the process of setting up GnoSwap contracts. The process involves three main steps: cloning the `gnoswap` repository, copying the contracts to the `gno` directory, and moving test cases to their respective directories.

To set up GnoSwap contracts in Gno Core, follow these steps:

1. Clone the `gnoswap` contracts repository:

   > Tip: If `$WORKDIR` is home directory, then `$WORKDIR` is `~/`.

   ```bash
   cd $WORKDIR
   git clone https://github.com/gnoswap-labs/gnoswap.git
   cd gnoswap
   ```

2. Copy the `gnoswap` contracts into the cloned `gno` repository:

   ```bash
   # copy grc20 tokens
   cp -R __local/grc20_tokens/* $WORKDIR/gno/examples/gno.land/r/demo
   cp -R _deploy/r/gnoswap $WORKDIR/gno/examples/gno.land/r/demo

   # copy gnoswap base packages ( includes uint256, int256 and bit of pool calculation )
   cp -R _deploy/p/gnoswap $WORKDIR/gno/examples/gno.land/p/demo

   # copy gnoswap base realms ( includes common logic, variables and consts )
   cp -R _deploy/r/gnoswap $WORKDIR/gno/examples/gno.land/r/demo/gnoswap

   # copy gnoswap realms
   cp -R pool position router staker $WORKDIR/gno/examples/gno.land/r/demo

>    ```

Reviewer: Just double checking: should we remove `demo` here?

Author: auto suggestion issue. `demo` should be removed

3. Move all test cases into its own directory:

Move the test cases for each contract to their respective directories. It's not necessary to move all tests; you can selectively move only the tests you need. However, files containing `VARS_HELPERS` in their name must be moved.

   ```bash
   # Pool
   cd $WORKDIR/gno/examples/gno.land/r/pool
   mv _TEST_/* .

   # Position
   cd $WORKDIR/gno/examples/gno.land/r/position
   mv _TEST_/* .

   # Router
   cd $WORKDIR/gno/examples/gno.land/r/router
   mv _TEST_/* .

   # Staker
   cd $WORKDIR/gno/examples/gno.land/r/staker
   mv _TEST_/* .

>    ```

Reviewer: shouldn't it be `gno.land/r/gnoswap/{name}`?

Author: https://github.com/gnoswap-labs/gnoswap/pull/288/commits/2a7b69ea864c1b12a8e9d51006c7761305dccf02

### Running Tests

While it's possible to run tests in the cloned `gno` directory (where the above setup process was completed), it's recommended to run them in the `gnoswap` directory to avoid confusion due to the large number of changed files.

First, navigate to the `gnoswap` directory:

```bash
cd gnoswap
```

Next, move to the Realm directory you want to test (such as `pool`, `staker`, etc.), then run the tests using the `gno test` command:

```bash
gno test -v . # or specify a particular test file path
```

## Realms

This section provides information about the core realms of GnoSwap that have been deployed.

### Core Realms Deployed on Testnet4

- pool: [gno.land/r/gnoswap/pool](https://gnoscan.io/realms/details?path=gno.land%2Fr%2Fgnoswap%2Fpool)
- position: [gno.land/r/gnoswap/position](https://gnoscan.io/realms/details?path=gno.land%2Fr%2Fgnoswap%2Fposition)
- router: [gno.land/r/gnoswap/router](https://gnoscan.io/realms/details?path=gno.land%2Fr%2Fgnoswap%2Frouter)
- staker: [gno.land/r/gnoswap/staker](https://gnoscan.io/realms/details?path=gno.land%2Fr%2Fgnoswap%2Fstaker)

### Pool

> Pool is a core component of Gnoswap, a smart contract that provides liquidity between two tokens and enables trading. Each pool has a unique token pair and fee tier, and operates based on Uniswap V3's concentrated liquidity mechanism.

Reviewer: ```suggestion
Pool is a core component of GnoSwap, a smart contract that facilitates liquidity provision and trading between two GRC20 tokens. Each pool has a unique token pair, fee tier, and customizable liquidity range, leveraging Uniswap V3's concentrated liquidity mechanism.
```

Key features:

>  - Composed of two tokens.

Reviewer: ```suggestion
 - Composed of two GRC20 tokens.
```

>  - Allows liquidity provision within a user-defined price range.

Reviewer: ```suggestion
 - Allows liquidity provision within a user-defined, customizable price range.
```

 - Supports various fee tiers, suitable for different trading strategies.
 - Dynamically adjusts liquidity according to price fluctuations.

### Position

> In GnoSwap, a Position is an NFT (non-fungible token) representing a liquidity provider's (LP's) unique liquidity token. Each position has the following key functions and properties:

Reviewer: ```suggestion
Position is a GRC721 NFT (non-fungible token) representing the liquidity provider's (LP's) unique liquidity position. Each position has the following key functions and properties:
```

> # Gnoswap Contracts

Reviewer: ```suggestion
# GnoSwap Contracts
```

This repository contains smart contracts (realms) for GnoSwap.

> ## Setting Up and Testing Gnoswap Contracts

Reviewer: ```suggestion
## Setting Up and Testing GnoSwap Contracts
```

1. Clone the `gnoswap` contracts repository:

   > Tip: If `$WORKDIR` is home directory, then `$WORKDIR` is `~/`.

   ```bash
   cd $WORKDIR
   git clone https://github.com/gnoswap-labs/gnoswap.git
   cd gnoswap
   ```

2. Copy the `gnoswap` contracts into the cloned `gno` repository:

   ```bash
   # copy grc20 tokens
   cp -R __local/grc20_tokens/* $WORKDIR/gno/examples/gno.land/r/demo
   cp -R _deploy/r/gnoswap $WORKDIR/gno/examples/gno.land/r/demo

   # copy gnoswap base packages ( includes uint256, int256 and bit of pool calculation )
   cp -R _deploy/p/gnoswap $WORKDIR/gno/examples/gno.land/p/demo

   # copy gnoswap base realms ( includes common logic, variables and consts )
   cp -R _deploy/r/gnoswap $WORKDIR/gno/examples/gno.land/r/demo/gnoswap

   # copy gnoswap realms
   cp -R pool position router staker $WORKDIR/gno/examples/gno.land/r/demo
   ```

3. Move all test cases into its own directory:

Move the test cases for each contract to their respective directories. It's not necessary to move all tests; you can selectively move only the tests you need. However, files containing `VARS_HELPERS` in their name must be moved.

   ```bash
   # Pool

>    cd $WORKDIR/gno/examples/gno.land/r/demo/pool

Reviewer: need to remove `demo`
